common/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,847 @@
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+
+namespace {
+
+
+template <typename ValueType>
+__device__ __forceinline__ void copy(
+    const gko::batch_dense::BatchEntry<const ValueType> &src_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &dest_shared_entry,
+    const uint32 &converged)
+{
+    for (int li = threadIdx.x;
+         li < dest_shared_entry.num_rhs * dest_shared_entry.num_rows;
+         li += blockDim.x) {
+        int r = li / dest_shared_entry.num_rhs;
+        int c = li % dest_shared_entry.num_rhs;
+
+        const uint32 conv = converged & (1 << c);
+
+        if (conv) {
+            continue;
+        }
+
+        dest_shared_entry.values[r * dest_shared_entry.stride + c] =
+            src_shared_entry.values[r * src_shared_entry.stride + c];
+    }
+}
+
+
+template <typename BatchMatrixType_entry, typename ValueType>
+__device__ __forceinline__ void initialize(
+    const gko::batch_dense::BatchEntry<ValueType> &rho_old_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &omega_old_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &alpha_shared_entry,
+    const BatchMatrixType_entry &A_global_entry,
+    const gko::batch_dense::BatchEntry<const ValueType> &b_global_entry,
+    const gko::batch_dense::BatchEntry<const ValueType> &x_global_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &x_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &r_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &r_hat_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &p_shared_entry,
+    const gko::batch_dense::BatchEntry<ValueType> &v_shared_entry,
+    const gko::batch_dense::BatchEntry<typename gko::remove_complex<ValueType>>
+        &rhs_norms_shared_entry,
+    const gko::batch_dense::BatchEntry<typename gko::remove_complex<ValueType>>
+        &res_norms_shared_entry)
+{
+    for (int c = threadIdx.x; c < rho_old_shared_entry.num_rhs;
+         c += blockDim.x) {
+        rho_old_shared_entry.values[0 * rho_old_shared_entry.stride + c] =
+            one<ValueType>();
+        omega_old_shared_entry.values[0 * omega_old_shared_entry.stride + c] =
+            one<ValueType>();
+        alpha_shared_entry.values[0 * alpha_shared_entry.stride + c] =
```

Comment:
I guess for readability you added the mutliplication by 0, but I think you can remove it.

Reply:
Removed.